### PR TITLE
L1-constrained regression using Frank-Wolfe

### PR DESCRIPTION
**Pull request by mblondel**
I implemented the FW method for L1-constrained regression. 
The method is greedy in nature : at most one non-zero coefficient is added at every iteration. I added an option to stop when a model size limit is reached.

Regularization paths of constrained (FW) and penalized (coordinate descent) formulations on the diabetes dataset:
fw

I still need to add docstrings and tests.

CC @fabianp @zermelozf @vene

### CODE REVIEW

**Comment by fabianp** [comment]
Cool ! I'll be offline until Monday, then I'll definitely take a look into it.

File: lightning/impl/fw.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            print "l1 norm", beta * np.sum(np.abs(w))
+            print "n_nz", np.sum(w != 0)
+            print
+
+        # Find optimal step size by exact line search.
+        Xs = s * X[:, atom]
+        if sp.issparse(Xs):
+            Xs_sq = np.dot(Xs.data, Xs.data)
+        else:
+            Xs_sq = np.dot(Xs, Xs)
```
Comment (vene) [inline lightning/impl/fw.py]: Since the sign doesn't affect this, you could precompute all column square norms outside of the loop, right?

File: lightning/impl/fw.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        y_pred_sq = np.dot(y_pred, y_pred)
+        b = (Xs - y_pred)
```
Comment (vene) [inline lightning/impl/fw.py]: b seems unused

File: lightning/impl/fw.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        gamma = np.dot(resid, y_pred) - safe_sparse_dot(resid, Xs)
+        gamma /= beta * (Xs_sq - 2 * safe_sparse_dot(Xs.T, y_pred) + y_pred_sq)
```
Comment (vene) [inline lightning/impl/fw.py]: When running the example I get a warning at this line, because Xs - y_pred is zero. The line below fixes gamma, but I thought it might be worth it to point it out.

File: lightning/impl/fw.py (hunk 1 above)
Comment (mblondel) [inline lightning/impl/fw.py]: Good idea, I'll do that. O(n_features) memory cache is not big deal.

File: lightning/impl/fw.py (hunk 3 above)
Comment (mblondel) [inline lightning/impl/fw.py]: Yep I get a warning too. I need to check more throughly what's the right thing to do in this case: set gamma=1, set gamma=0 or stop the algorithm?

**Comment by fabianp** [comment]
since it is specific to least squares I would call it _frank_wolfe_ls (and also because I would love a FW for arbitrary loss functions :-). 

Just FYI yesterday I saw this paper (http://arxiv.org/pdf/1511.05932.pdf), in which they prove that a small modification to FW yield an algorithm with linear convergence rate, although I don't know whether in practice this has always a major effect.

**Comment by mblondel** [comment]
> since it is specific to least squares I would call it _frank_wolfe_ls (and also because I would love a FW for arbitrary loss functions :-)

This shouldn't be too hard. The only difficulty I see is for computing the step size. For arbitrary loss it's not possible to compute the exact step size but a few iterations of one-dimensional Newton method should work.

**Comment by fabianp** [comment]
What about using [scipy's line_search](http://docs.scipy.org/doc/scipy-0.16.0/reference/generated/scipy.optimize.line_search.html) for arbitrary loss functions and the exact one for the quadratic one?

**Comment by mblondel** [comment]
Not sure, does it work for constrained optimization?

**Comment by fabianp** [comment]
you are right, probably not

**Comment by vene** [comment]
I was just reading on [hybrid conditional gradient - smoothing](http://arxiv.org/abs/1404.3591) and it seems like it could lead to an efficient way to extend this with an additional group lasso penalty.

**Comment by mblondel** [comment]
@vene @fabianp For Frank-Wolfe, the regularization path matches the one of the penalized version trained by coordinate descent, as expected. However, for FISTA with `penalty=l1-ball`, it doesn't. Either FISTA is not supposed to handle constrained problems or there is a problem with my implementation. I am observing issues with constrained problems with [this implementation](https://gist.github.com/mblondel/5105786d740693a6996bcb8e482c7083) too.

**Comment by fabianp** [comment]
In theory FISTA should work with constrained problems using the projection as proximal operator. Have you tried with simple ISTA=projected gradient descent to debug (trivial to implement)?

**Comment by vene** [comment]
[Bach et al (p12 here)](http://lear.inrialpes.fr/people/mairal/resources/pdf/opt_book.pdf) say that "proximal methods apply" and don't seem to suggest the need for any special treatment.

**Comment by mblondel** [comment]
There was an issue in `project_simplex`, see dfb8586d3069beaaef0bcc3d1769fa8316d675f2.

With a small tweak to the plot code, FISTA is now looking fine.
